File: hackerrank/calculate-gcd/calculate_gcd.rb
# frozen_string_literal: true

puts 'Running in ruby'

def min(first, second)
  first < second ? first : second
end

def max(first, second)
  first > second ? first : second
end

def gcd(first, second)
  first, second = min(first, second), max(first, second)

  first, second = second % first, first until (second % first).zero?

  first
end

def gcd_multiple(*numbers)
  numbers.reduce { |first, second| gcd(first, second) }
end

(0...gets.to_i).each do |_|
  numbers = gets.split.map(&:to_i)
  puts "gcd(#{numbers.join(', ')}) = #{gcd_multiple(*numbers)}"
end

File: hackerrank/calculate-gcd/calculate-gcd.py
#!/usr/bin/env python

from functools import reduce

print("Running in python")

def gcd(first, second):
  first, second = min(first, second), max(first, second)

  while second % first != 0:
    first, second = second % first, first

  return first

def gcd_multiple(*numbers):
  return reduce(lambda first, second: gcd(first, second), numbers)


for t in range(0, int(input())):
  numbers = list(map(int, input().split(' ')))
  print("gcd(%s) = %s" % (', '.join(map(str, numbers)), str(gcd_multiple(*numbers))))
